### PR TITLE
mongo: initial implementation

### PR DESCRIPTION
**Pull request by Altonhe**
### Add support for MongoDB

add support for MongoDB

### Checklist

- [x] I have added test cases to cover the new code.
- [x] I agree to follow this project's [Code of Conduct](https://golang.org/conduct) by submitting this pull request.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -9,4 +9,5 @@ require (
 	github.com/jackc/pgx/v4 v4.11.0
 	github.com/pkg/errors v0.9.1
 	github.com/stretchr/testify v1.7.0
+	go.mongodb.org/mongo-driver v1.7.3 // indirect
```
Comment (unknwon) [inline go.mod:12]: I think the `// indirect` part will be removed if you run `go mod tidy`. This appears to be a direct dependency not an indirect dependency.

File: mongo/mongo.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+	"github.com/flamego/session"
+)
+
+var _ session.Store = (*mongoStore)(nil)
+
+// mongoStore is a MongoDB implementation of the session store.
+type mongoStore struct {
+	nowFunc    func() time.Time // The function to return the current time
+	lifetime   time.Duration    // The duration to have no access to a session before being recycled
```
Comment (unknwon) [inline mongo/mongo.go:26]:
```suggestion
	lifetime   time.Duration    // The duration to have access to a session before being recycled
```

This looks like a typo in the original copy, could you also update the comment in `mysql/mysql.go` and `postgres/postgres.go`?

File: mongo/mongo.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+		db:         cfg.db,
+		collection: cfg.Collection,
+		encoder:    cfg.Encoder,
+		decoder:    cfg.Decoder,
+	}
+}
+
+func (s mongoStore) Exist(ctx context.Context, sid string) bool {
+	err := s.db.Collection(s.collection).FindOne(ctx, bson.M{"key": sid}).Err()
+	return err != mongo.ErrNoDocuments
```
Comment (unknwon) [inline mongo/mongo.go:47]: I think this is not accurate, we should treat any error as not exists, so
```suggestion
	return err == nil
```

File: mongo/mongo.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+		sess := session.NewBaseSession(sid, s.encoder)
+		sess.SetData(data)
+		return sess, nil
+	} else if err != mongo.ErrNoDocuments {
+		return nil, errors.Wrap(err, "select")
+	}
+
+	return session.NewBaseSession(sid, s.encoder), nil
+
```
Comment (unknwon) [inline mongo/mongo.go:82]:
```suggestion
```

File: mongo/mongo.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	return err != mongo.ErrNoDocuments
+}
+
+func (s mongoStore) Read(ctx context.Context, sid string) (session.Session, error) {
+	var result bson.M
+	err := s.db.Collection(s.collection).FindOne(ctx, bson.M{"key": sid}).Decode(&result)
+	if err == nil {
+		binary, ok := result["data"].(primitive.Binary)
+		if !ok {
+			return nil, errors.New("assert `data` key")
```
Comment (unknwon) [inline mongo/mongo.go:56]:
```suggestion
			return nil, errors.Errorf(`assert "data" key: want type primitive.Binary but got %T`, result["data"])
```

I think this is more common to use double quotes to quota a key name.

File: mongo/mongo.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		}
+
+		expiredAt, ok := result["expired_at"].(primitive.DateTime)
```
Comment (unknwon) [inline mongo/mongo.go:59]:
```suggestion
		expiredAt, ok := result["expiredAt"].(primitive.DateTime)
```

Is MongoDB's naming convention is to use snake case? If so, please ignore this suggestion.

File: mongo/mongo.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+		if !ok {
+			return nil, errors.New("assert `expired_at` key")
```
Comment (unknwon) [inline mongo/mongo.go:61]:
```suggestion
			return nil, errors.Errorf(`assert "expired_at" key: want type primitive.DateTime but got %T`, result["expired_at"])
```

File: mongo/mongo.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+		nowFunc:    cfg.nowFunc,
+		lifetime:   cfg.Lifetime,
+		db:         cfg.db,
+		collection: cfg.Collection,
+		encoder:    cfg.Encoder,
+		decoder:    cfg.Decoder,
+	}
+}
+
+func (s mongoStore) Exist(ctx context.Context, sid string) bool {
```
Comment (unknwon) [inline mongo/mongo.go:45]:
```suggestion
func (s *mongoStore) Exist(ctx context.Context, sid string) bool {
```

What is the reason that we do not use pointer receiver here? Same for all other methods.

File: mongo/mongo.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+			Upsert: &upsert,
+		})
+	if err != nil {
+		return errors.Wrap(err, "upsert")
+	}
+	return nil
+}
+
+func (s mongoStore) GC(ctx context.Context) error {
+	_, err := s.db.Collection(s.collection).DeleteMany(ctx, bson.M{"expired_at": bson.M{"$lt": s.nowFunc().UTC()}})
```
Comment (unknwon) [inline mongo/mongo.go:115]: We should use less than or equal to, not just the former.

File: mongo/mongo.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+		data, err := s.decoder(binary.Data)
+		if err != nil {
+			return nil, errors.Wrap(err, "decode")
+		}
+
+		sess := session.NewBaseSession(sid, s.encoder)
+		sess.SetData(data)
+		return sess, nil
+	} else if err != mongo.ErrNoDocuments {
+		return nil, errors.Wrap(err, "select")
```
Comment (unknwon) [inline mongo/mongo.go:78]:
```suggestion
		return nil, errors.Wrap(err, "find")
```

File: mongo/mongo.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+	if err != nil {
+		return errors.Wrap(err, "upsert")
+	}
+	return nil
+}
+
+func (s mongoStore) GC(ctx context.Context) error {
+	_, err := s.db.Collection(s.collection).DeleteMany(ctx, bson.M{"expired_at": bson.M{"$lt": s.nowFunc().UTC()}})
+	if err != nil {
+		return errors.Wrap(err, "GC")
```
Comment (unknwon) [inline mongo/mongo.go:117]:
```suggestion
		return errors.Wrap(err, "delete")
```

File: mongo/mongo.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+// Options keeps the settings to set up Redis client connection.
+type Options = options.ClientOptions
+
+// Config contains options for the Redis session store.
+type Config struct {
+	// For tests only
+	nowFunc func() time.Time
+	db      *mongo.Database
+
+	// Options is the settings to set up MongoDB client connection.
```
Comment (unknwon) [inline mongo/mongo.go:131]:
```suggestion
	// Options is the settings to set up the MongoDB client connection.
```

File: mongo/mongo.go [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+			cfg.nowFunc = time.Now
+		}
+		if cfg.Lifetime.Seconds() < 1 {
+			cfg.Lifetime = 3600 * time.Second
+		}
+		if cfg.Collection == "" {
+			cfg.Collection = "sessions"
+		}
+		if cfg.Encoder == nil {
+			cfg.Encoder = session.GobEncoder
```
Comment (unknwon) [inline mongo/mongo.go:181]: Shouldn't `bson.Encoder` be used here (as said in docstring)?

File: mongo/mongo.go [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+// Config contains options for the Redis session store.
+type Config struct {
+	// For tests only
+	nowFunc func() time.Time
+	db      *mongo.Database
+
+	// Options is the settings to set up MongoDB client connection.
+	Options *Options
+	// DSN is the database source name to the MongoDB.
+	DSN string
```
Comment (unknwon) [inline mongo/mongo.go:134]:
```suggestion
	// Database is the database name of the MongoDB.
	Database string
```

Based on my understanding, this field is purely used as a database name, not as a DSN.

File: mongo/mongo.go [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+			cfg.Lifetime = 3600 * time.Second
+		}
+		if cfg.Collection == "" {
+			cfg.Collection = "sessions"
+		}
+		if cfg.Encoder == nil {
+			cfg.Encoder = session.GobEncoder
+		}
+		if cfg.Decoder == nil {
+			cfg.Decoder = session.GobDecoder
```
Comment (unknwon) [inline mongo/mongo.go:184]: Ditto

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for the PR!

I haven't finished reviewing all the files but submitting my current review for changes requested.

File: go.mod (hunk 1 above)
Comment (Altonhe) [reply go.mod:12]: Exactly.

File: mongo/mongo.go (hunk 6 above)
Comment (Altonhe) [reply mongo/mongo.go:59]: It looks like in the official example they're using snake case.
https://sourcegraph.com/github.com/mongodb/docs-golang@master/-/blob/source/includes/usage-examples/code-snippets/updateOne.go?L40

File: mongo/mongo.go (hunk 9 above)
Comment (Altonhe) [reply mongo/mongo.go:114]: Exactly.

File: mongo/mongo.go (hunk 13 above)
Comment (Altonhe) [reply mongo/mongo.go:180]: The docstring here is not correct, we should keep using `session.GobEncoder`.

File: mongo/mongo.go (hunk 15 above)
Comment (Altonhe) [reply mongo/mongo.go:183]: Same as above.

File: mongo/mongo.go (hunk 2 above)
Comment (Altonhe) [reply mongo/mongo.go]: Sure.

File: mongo/mongo.go (hunk 6 above)
Comment (unknwon) [reply mongo/mongo.go:59]: Cool, then let's stick with snake case.

File: mongo/mongo_test.go [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+		err = db.Collection("sessions").Drop(ctx)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func TestMongoDBStore(t *testing.T) {
```
Comment (unknwon) [inline mongo/mongo_test.go:61]:
```suggestion
func TestMongoStore(t *testing.T) {
```

We have been calling it `MongoStore` in `mongo.go`.

File: mongo/mongo_test.go [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+	resp = httptest.NewRecorder()
+	req, err = http.NewRequest(http.MethodGet, "/destroy", nil)
+	assert.NoError(t, err)
+
+	req.Header.Set("Cookie", cookie)
+	f.ServeHTTP(resp, req)
+	assert.Equal(t, http.StatusOK, resp.Code)
+}
+
+func TestMongoDBStore_GC(t *testing.T) {
```
Comment (unknwon) [inline mongo/mongo_test.go:129]:
```suggestion
func TestMongoStore_GC(t *testing.T) {
```

Ditto.

**Review by unknwon — APPROVED** [review]
Thanks for the follow up!

Two last suggestions, approve to unblock.

**Pull request merged**